liberty-maven-plugin/src/main/java/net/wasdev/wlp/maven/plugins/server/DevMojo.java

```diff
@@ -0,0 +1,889 @@
+                        log.info("Changes were detected while tests were running. Restarting tests.");
+                    }
+                    return;
+                }
+
+                if (!skipITs) {
+                    if (waitForApplicationUpdate) {
+                        // wait until application has been updated
+                        if (appUpdateTimeout < 0) {
+                            appUpdateTimeout = 5;
+                        }
+                        long timeout = appUpdateTimeout * 1000;
+                        serverTask.waitForUpdatedStringInLog(regexp, timeout, logFile, messageOccurrences);
+                    }
+
+                    log.info("Running integration tests...");
+                    try {
+                        runIntegrationTests();
+                        log.info("Integration tests finished.");
+                    } catch (MojoExecutionException e) {
+                        Throwable cause = e.getCause();
+                        if (cause != null && cause instanceof MojoFailureException) {
+                            log.debug(e);
+                            log.error("Integration tests failed: " + cause.getLocalizedMessage());
+                        } else {
+                            log.error("Failed to run integration tests", e);
+                        }
+                    }
+                }
+            }
+
+            // finally, start watching for hotkey presses if not already started
+            util.runHotkeyReaderThread(executor);
+        }
+
+        @Override
+        public void checkConfigFile(File configFile, File serverDir) {
+            try {
+                ServerFeature servUtil = new ServerFeature();
+                Set<String> features = servUtil.getServerFeatures(serverDir);
+                features.removeAll(existingFeatures);
+                if (!features.isEmpty()) {
+                    List<String> configFeatures = new ArrayList<String>(features);
+                    log.info("Configuration features have been added");
+                    runMojo("net.wasdev.wlp.maven.plugins:liberty-maven-plugin", "install-feature", serverName,
+                            configFeatures);
+                    this.existingFeatures.addAll(features);
+                }
+            } catch (Exception e) {
+                log.debug("Failed to read configuration file", e);
+            }
+        }
+
+        @Override
+        public boolean compile(File dir) {
+            try {
+                if (dir.equals(sourceDirectory)) {
+                    log.info("Running maven-compiler-plugin:compile");
+                    runMojo("org.apache.maven.plugins:maven-compiler-plugin", "compile", null, null);
+
+                    log.info("Running maven-compiler-plugin:resources");
+                    runMojo("org.apache.maven.plugins:maven-resources-plugin", "resources", null, null);
+                }
+                if (dir.equals(testSourceDirectory)) {
```

Comment:
This is a nit, but this could be an 'else if'.

Reply:
sourceDirectory and testSourceDirectory could possibly be the same directory (even though that is bad practice and probably no one does that).